### PR TITLE
pkg/emlearn: add support to RIOT

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adds support for the [emlearn](https://github.com/emlearn/emlearn) machine learning inference library as a package.

The library allows to use a C header file that contains the model structure and the associated weights. It also supports different classifiers, not only neural networks.
For example, the provided test application shows how to train a [Scikit-Learn](https://scikit-learn.org) Random Forest estimator to predict the value of a hand-written digit in a an image.

### Testing procedure

The following command shoud work:
```
make BOARD=<board of your choice> -C tests/pkg_emlearn flash test
```

### Issues/PRs references

None

### CODE REVIEW

**Comment by aabadie** [comment]
I just realized that for the test to work, the docker container must contain the scikit-learn, joblib and emlearn Python packages.

**Comment by aabadie** [comment]
This will probably fail on Murdock because it lacks some Python requirements (scikit-learn, emlearn). Or maybe the new docker image is already deployed on the workers ?

Let me postpone the CI build until we have a confirmation.

**Comment by benpicco** [comment]
Looks like Murdock is up to date now, but the example is a bit too hefty for some boards.

**Comment by aabadie** [comment]
> the example is a bit too hefty for some boards.

This is fixed in b077b87 @benpicco.

**Comment by aabadie** [comment]
The test triggers a hard fault on hardware, I'll have a look...

**Comment by benpicco** [comment]
Well no more crashes 😄 
Only `run_test/examples/suit_update/samr21-xpro:gnu` fails, but that's unrelated.

File: tests/pkg_emlearn/Makefile [hunk 1]
```diff
@@ -0,0 +1,22 @@
+include ../Makefile.tests_common
+
+USEPKG += emlearn
+
+BLOBS += digit
+
+BUILDDEPS += sonar.h
+
+# Disable hardware FPU for Cortex-M4/7 to prevent crash on these platforms
+DISABLE_MODULE += cortexm_fpu
```
Comment (benpicco) [inline tests/pkg_emlearn/Makefile]: Is this really needed?
Before we've seen crashes on `samr21-xpro` which is a cm0+ and doesn't have a FPU.

I've tried this now without the `DISABLE_MODULE += cortexm_fpu` on both  `samr21-xpro` and `same54-xpro` (cm4f) and it worked just fine.

**Comment by benpicco** [comment]
I think this could be merged, but I'd prefer if we could do without disabling `cortexm_fpu`.

**Comment by aabadie** [comment]
> I'd prefer if we could do without disabling cortexm_fpu.

I don't think that's possible: it crashes on all cpus that provide an FPU if it's enabled. It doesn't if it's not. I tried nrf52832, stm32f429i-disc1, stm32f723e-disco, nucleo-l476rg.
The main issue for me is that it also crashes on cortexm0. I tried samr21-xpro, samr34-xpro and arduino-mkr1000. For the moment, there's no simple way to automatically blacklist those boards based on their cpu core model (cortexm0 here). The only that could be done is to explicitly blacklist them, but I'd like to avoid that.

**Comment by benpicco** [comment]
Looks like this is not a problem with the FPU itself but with the unaligned `digit` array.

With this patch the test runs without crashing on `same54-xpro`:

```patch
diff --git a/makefiles/blob.inc.mk b/makefiles/blob.inc.mk
index c907fbe8b1..cfc963c587 100644
--- a/makefiles/blob.inc.mk
+++ b/makefiles/blob.inc.mk
@@ -58,7 +58,8 @@ $(BLOB_H): $(BLOB_HDR_DIR)/%.h: % $(BLOBS) | $$(@D)/.
        @command -v $(XXD) > /dev/null 2>&1 || ( \
          echo "error: xxd binary '$(XXD)' not found! (Maybe install vim package?)"; exit 1)
 
-       ${Q}cd $(dir $<); xxd -i $(notdir $<) | sed 's/^unsigned/const unsigned/g'> $@
+       ${Q}cd $(dir $<); xxd -i $(notdir $<) | sed 's/^unsigned/const unsigned/g' | \
+       sed '1iconst unsigned char $(basename $<)[] __attribute__ ((aligned (4)));' > $@
 
 # make C and C++ objects of this module depend on generated headers, so they
 # get re-build on changes to the blob files.
```

even when I remove the `DISABLE_MODULE += cortexm_fpu`.

**Comment by aabadie** [comment]
> Looks like this is not a problem with the FPU itself but with the unaligned digit array.

Oh, thanks a lot ! I was pretty sure this package has worked before on samr21-xpro and other M4 but couldn't find what introduced the issue.

**Comment by aabadie** [comment]
@benpicco, I pushed a couple of commits that are working without having to patch the blobs mechanism. This is a bit ugly, but, somehow, I like it :)

File: tests/pkg_emlearn/main.c [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * @}
+ */
+
+#include <stdio.h>
+#include <inttypes.h>
+
+#include "model.h"
+
+/* the digit array included must be 4-byte aligned */
+__attribute__((__aligned__(4)))
```
Comment (benpicco) [inline tests/pkg_emlearn/main.c:27]: Hah, I didn't know this was possible.

**Comment by benpicco** [comment]
Works on `same54-xpro` and on `samr21-xpro` - please squash.

File: tests/pkg_emlearn/main.c (hunk 2 above)
Comment (aabadie) [reply tests/pkg_emlearn/main.c:27]: With the `#include` statement, the preprocessor just replace the line with the content of the file. So in the end, the "generated" code is correct. It's a bit hack :)

**Comment by benpicco** [comment]
Failed tests were

```
    tests/pkg_emlearn/nrf52dk:gnu
    tests/thread_msg_block_race/nrf52dk:llvm
    tests/xtimer_now64_continuity/nrf52dk:llvm
    examples/micropython/nrf52dk:gnu
    tests/cpp11_condition_variable/nrf52dk:gnu
    tests/pkg_libfixmath/nrf52dk:gnu
    tests/pkg_micro-ecc/nrf52dk:gnu
    tests/posix_time/nrf52dk:llvm
    tests/pthread/nrf52dk:gnu
    tests/netdev_test/nrf52dk:llvm
    tests/pthread_rwlock/nrf52dk:gnu
    tests/evtimer_msg/nrf52dk:gnu
    tests/shell/nrf52dk:llvm
    tests/vfs_plus_stdio/nrf52dk:llvm
    tests/gnrc_sock_neterr/nrf52dk:llvm
    tests/msg_avail/nrf52dk:llvm
    tests/irq_disable_restore/nrf52dk:llvm
```

looks like there is something wrong with `nrf52dk` on CI. No log for these, just 404.

```
    tests/pkg_emlearn/nrf52dk:llvm
    tests/pkg_emlearn/samr21-xpro:llvm
    tests/pkg_emlearn/esp32-wroom-32:gnu
    tests/pkg_emlearn/samr21-xpro:gnu
```

all passed.

**Pull request merged**